Add Secrets Manager interface [Draft]

wjordan opened the pull request:
This PR contains a first pass at integrating Secrets Manager secrets into the CDO namespace of our application.

- Add `aws-sdk-secretsmanager` gem
- Add `secretsmanager` IAM permissions to stack
- Add `Cdo::Secrets` interface for fetching secrets

### Update (May 2019)
I substantially rewrote/refined the implementation incorporating feedback so it works differently from the original description- this is the updated summary:

The `Cdo::Secrets` instance contains an `Aws::SecretsManager::Client` connection, which provides a `#get` method that asynchronously (using [`Promises`](http://ruby-concurrency.github.io/concurrent-ruby/master/Concurrent/Promises.html)) fetches individual secrets using the [`GetSecretValue`](https://docs.aws.amazon.com/secretsmanager/latest/apireference/API_GetSecretValue.html) API. A `#lazy` method allows a secret to be lazy-loaded (asynchronously fetch in the background, then wait for fetch to complete when first used), and a `#required!` method will wait until _all_ referenced secrets are fully loaded (and will raise an error if any fail to load).

A `CDO.secrets` global variable is set to an instance of `Cdo::Secrets`.

Secrets can be sequentially loaded from a set of path-prefixes (currently, `[environment]/cdo/` followed by `shared/cdo/`) that serve as namespaces. This will allow us to place a secret either in `shared/cdo/` to share across all CDO applications, or in e.g., `production/cdo/` to share in a specific environment, to support the role+environment-based secrets loading currently being used for Chef-attribute secrets.

The recommended way to reference Secrets will be through YAML configuration. The `Cdo::Secret` class implements a new custom YAML tag `!Secret`, which indicates a Secret value that will be lazy-loaded from the provided environment, defaulting to the same key as the YAML key.

This example will fetch `secret` from configured path-prefixes (e.g., it will check `[environment]/cdo/secret` then `shared/cdo/secret`, in that order):

```yaml
secret: !Secret
```

The tag takes some optional values to specify an optional key or fallback string, e.g.:

```yaml
secret2: !Secret z # Fetch `[environment]/cdo/z` or `shared/cdo/z`.
secret3: !Secret x/y/z # Fetch fully-qualified `x/y/z`.
secret4: !Secret [missing_key, 'not a secret!'] # Fetch `missing_key`, or `'not a secret!'` if not found.
secret5: !Secret [x/y/z, 'not a secret!'] # Fetch fully-qualified `x/y/z`, or `'not a secret!'` if not found.
```

In this iteration, creating and updating secrets will be done directly from the AWS Console (or directly through the API). This is still a bit cumbersome- future iterations will consider more developer-friendly processes for managing secrets, and will also consider approaches to automated rotating of existing secrets.

uponthesun commented on lib/cdo/secrets.rb:
```diff
@@ -0,0 +1,61 @@
+
+      # Asynchronously load {name => id} Hash of secrets from provided path prefixes.
+      @secret_names = Concurrent::Promises.future do
+        @client = client || Aws::SecretsManager::Client.new
```
nit: does initializing the client need to be in the asynchronous block? If not, I'd prefer to move it outside / before it.

wjordan replied:
Initializing the Client also initializes credentials which can take a non-trivial time in some configurations (e.g., [Instance Profile](https://github.com/aws/aws-sdk-ruby/blob/master/gems/aws-sdk-core/lib/aws-sdk-core/instance_profile_credentials.rb) or [Process](https://github.com/aws/aws-sdk-ruby/blob/master/gems/aws-sdk-core/lib/aws-sdk-core/process_credentials.rb) credentials), which is why I included them in the Promise.

I agree it's not ideal to include client initialization in the @secret_names promise though. I'll see if I can refactor this (maybe move client init into a separate promise).

uponthesun replied:
I'm fine with any of these options, including leaving as-is.

wjordan replied:
Refactored client-init to a separate promise in 01d6f49bac16f245bb84fb5e564e2ea58a099ad6.

uponthesun commented on lib/cdo/secrets.rb:
```diff
@@ -0,0 +1,61 @@
+            to_h
+        end.inject(&:merge)
+      rescue Aws::Errors::MissingCredentialsError, Aws::SecretsManager::Errors::ServiceError
+        # Silently fail in development where AWS credentials may not be present.
```
What do you expect the usage pattern to be for secrets while developing locally? Should we make the two different modes more explicit, perhaps with a CDO.use_secrets value which defaults to true, but is set to false in locals.yml.default?

If you want to keep this as-is, I'd still recommend not swallowing ServiceError in development, as that could be very confusing when trying to use secrets in development.

wjordan replied:
you read my mind, I just tweaked this to depend on a `CDO.secrets_required` value so it can be locally configurable.

I think I'd still prefer to default to `false` for `development` though, so that it doesn't add any friction to the case where AWS/SecretsManager credentials aren't being used.

uponthesun commented on lib/cdo/secrets.rb:
```diff
@@ -0,0 +1,61 @@
+      # Asynchronously load {name => id} Hash of secrets from provided path prefixes.
+      @secret_names = Concurrent::Promises.future do
+        @client = client || Aws::SecretsManager::Client.new
+        secret_list = @client.list_secrets.secret_list.map(&:name)
```
Isn't this a paginated API? Don't you need to handle the case where there are more results?

wjordan replied:
The Ruby SDK actually handles pagination automatically- [`Aws::SecretsManager::Client`](https://github.com/aws/aws-sdk-ruby/blob/6b5a6f6838a0f61b8af26d443434554921dbab86/gems/aws-sdk-secretsmanager/lib/aws-sdk-secretsmanager/client.rb#L52-L52) adds the [`Aws::Plugins::ResponsePaging`](https://github.com/aws/aws-sdk-ruby/blob/6b5a6f6838a0f61b8af26d443434554921dbab86/gems/aws-sdk-core/lib/aws-sdk-core/plugins/response_paging.rb) plugin, which [includes](https://github.com/aws/aws-sdk-ruby/blob/6b5a6f6838a0f61b8af26d443434554921dbab86/gems/aws-sdk-core/lib/aws-sdk-core/plugins/response_paging.rb#L11) [`PageableResponse`](https://github.com/aws/aws-sdk-ruby/blob/master/gems/aws-sdk-core/lib/aws-sdk-core/pageable_response.rb) and sets its [`Pager`](https://github.com/aws/aws-sdk-ruby/blob/master/gems/aws-sdk-core/lib/aws-sdk-core/pager.rb) to the value [defined by the `SecretsManager::Client`](https://github.com/aws/aws-sdk-ruby/blob/master/gems/aws-sdk-secretsmanager/lib/aws-sdk-secretsmanager/client_api.rb#L472-L478):

```
        o[:pager] = Aws::Pager.new(
          limit_key: "max_results",
          tokens: {
            "next_token" => "next_token"
          }
        )
      end)
```

...(which fwiw, is auto-generated from an [API model document](https://github.com/aws/aws-sdk-ruby/blob/master/apis/secretsmanager/2017-10-17/paginators-1.json))

Put this all together and it gives you an [Enumerable](https://ruby-doc.org/core/Enumerable.html) that automatically pages until the end when you iterate on it.

uponthesun replied:
Huh, that's pretty nice. Do you think the comment on the SDK docs about NextToken is out of date / inaccurate? https://docs.aws.amazon.com/sdk-for-ruby/v3/api/Aws/SecretsManager/Client.html#list_secrets-instance_method

wjordan replied:
The SDK client documentation is auto-generated from the [API model](https://github.com/aws/aws-sdk-ruby/blob/master/apis/secretsmanager/2017-10-17/docs-2.json#L14), so the comment is accurate for the API itself at least. The SDK doesn't manually rewrite any of the documentation to match the effects of its plugins, so I wouldn't expect them to change that anytime soon.

wjordan commented:
Thanks for the feedback Winter, here are my responses to your general questions/comments:

> If I'm understanding correctly, this lazily retrieves the secret the first time it's needed, right? I'd rather secrets misconfiguration errors fail-fast on deployment. It could be something like - we have an explicitly configured list of all secrets we care about, and we try to retrieve all of them on startup. And then our code can only access keys listed this configuration. This seems similar to what you have implemented, except your current implementation uses the configured secrets as the authoritative list.

Yeah, I agree some kind of pre-fetching logic would be a useful (even necessary) addition, I'll work through this next. There are a few details I need to work through if we want eager-loadiang to scale up to a large number of secrets / app-server instances without straining API-limits or boot-times:

1. Smaller/lightweight processes (e.g., Rails console, cron tasks, UI-test runners) shouldn't eager-load all secrets if they only need one or two of them to do their thing.
    - Or maybe they shouldn't even call `ListSecrets` at all if a process might not need any secrets?
2. Eager-loaded secrets should be loaded in parallel rather than sequentially to reduce total boot time.
3. The list of secrets to eager-load could either be hard-coded in the application code (where we would need to keep it in sync with uses of the secrets), or it could be derived from the list returned by `ListSecrets`.

> How is this singleton object shared across worker threads / processes? Will each thread end up with its own copy of this data?

Threads in the same process reference the same in-memory storage (which synchronizes reads/writes so is thread-safe); Worker-processes across the same app-server each get a copy of secrets after preload initializers complete (when the worker-processes fork), any secrets loaded _after_ the preload-fork will be specific to a process.
So the more secrets are loaded pre-fork, the less duplicate API calls will be needed to fetch them.

> How does the caching behavior work? (From what I gather we're using something we wrote which is built on top of Rails' own caching interfaces, but I don't know how any of it works / is configured - any links which point to that would be helpful.)

`CacheMethod` is just a simple in-memory method-memoizing decorator I wrote, which uses `CDO.cache` as its cache-store by default:

https://github.com/code-dot-org/code-dot-org/blob/8b83139672c01539b7d3966035c6cf00dc83eaf2/lib/cdo/cache_method.rb#L7-L26

`CDO.cache` is implemented as an instance of [ActiveSupport::Cache::MemoryStore](https://github.com/rails/rails/blob/master/activesupport/lib/active_support/cache/memory_store.rb) (which implements the [`ActiveSupport::Cache::Store`](https://github.com/rails/rails/blob/master/activesupport/lib/active_support/cache.rb) interface):

https://github.com/code-dot-org/code-dot-org/blob/8b83139672c01539b7d3966035c6cf00dc83eaf2/lib/cdo/cache.rb#L4-L12

wjordan commented:
Updated the PR with a substantial refactor. In some ways it's more simple, in other ways more complex:

- Now only uses the `GetSecretValue` API to fetch specific secrets by name directly, so `ListSecrets` is no longer used/needed.
- Added a YAML tag `!Secret` for specifying secrets within config files (`locals.yml` / `globals.yml`)
  - The goal with this tag was a compromise to allow secrets to be usable for development environments, while at the same time not locking in _all_ application developers to an unavoidable dependency on AWS/Secrets Manager access.
- Secrets can be lazy-loaded, so they start fetching in the background and only block when they are actually used/required
  - This was to allow specifying secrets within the global configs used throughout the application (via `deployment.rb`, while not having every single script load _all_ secrets on every run.
- Added a `before_fork` app-server hook so all required secrets (including all `!Secret` tags specified in yaml config) are eager-loaded before fork and exception raised if any required secrets are not found.
  - This was to ensure that in the production app-server use case, a server doesn't finish starting unless all secrets it expects to be present are fully loaded, so we can ensure that there are no additional secrets-manager API calls required after the app-server starts serving requests.

uponthesun commented on lib/cdo/secrets.rb:
```diff
@@ -0,0 +1,129 @@
+
+    # Ensure all required secrets are fully loaded.
+    # @return [Hash] All keys and their resolved values.
+    def required!(*keys)
```
Can you add unit tests for this method?

wjordan replied:
added in 4c6ae31dc366d18befacbcc8ca7f5865960eb0c5

uponthesun commented on lib/cdo/secrets.rb:
```diff
@@ -0,0 +1,129 @@
+        client.then do |c|
+          # Call GetSecretValue within each path, returning the first result.
+          start = Concurrent::Promises.rejected_future(nil, @pool)
+          paths.inject(start) do |promise, path|
```
It feels like the use of Promises here is making this code unnecessarily complicated. Would it be possible to instead write a helper method `get_sync` or similar which tries all the paths in a synchronous way using simpler syntax, and then return a Promise which wraps a call to `get sync`?

wjordan replied:
Yeah, I originally thought that I needed to build all of the asynchronous conditional logic out of Promises, but you're right it was just complicating things- I've committed an update that's much simpler.

uponthesun commented on lib/cdo/lazy.rb:
```diff
@@ -0,0 +1,26 @@
+  # loaded!
+  # Lazy
+  # => nil
+  class Lazy < SimpleDelegator
```
Even though it's implicitly covered already, it's be worth adding dedicated unit tests for this class if it's intended to be reusable throughout the code base.

wjordan replied:
added in 198e3e5be73aae2c9f0f1efdcc02e8e31385afd6

wjordan commented:
> * searching multiple paths to resolve secrets (e.g. /env/cdo and /shared/cdo)

The goal of this feature is to support the secrets-resolution logic we currently depend on in Chef (via [Attribute Precedence](https://docs.chef.io/attributes.html#attribute-precedence)), where we set some shared secrets in a shared `baseline` Role, and optionally override them in specific Environments.

> * allowing both unqualified and fully-qualified secret names and implicitly choosing between them based on the presence of '/'

With the first feature in mind, I thought we might have cases where we want to reference the fully-qualified path in some cases, e.g., to ensure a secret is shared without any environment-specific override and avoid the second `GetSecretValue` API call. This was speculative though so I could probably remove support for this mode until we have a specific use-case that requires it, since nothing similar exists in our current Chef-attributes setup.

uponthesun commented:
"The goal of this feature is to support the secrets-resolution logic we currently depend on in Chef" - yea, that makes sense. It'd be nice to be able to do things in a way so that you don't have to know rules of inheritance / precedence to know where a secret comes from, but we might have too many secret values  x environments, and might need some features like this. 

Given that we have this logic, I think it'd be useful to log the resolved secret that's used for each key during loading. It serves as its own form of "documentation" for people who are learning the system.

wjordan commented:
Closing this draft PR; will reopen a new PR with an updated description of the final revision of this feature.
